libs/platform/ebpf_ring_buffer.c

```diff
@@ -0,0 +1,242 @@
+    ebpf_ring_descriptor_t* ring_descriptor;
+} ebpf_ring_buffer_t;
+
+inline static size_t
+_ring_get_producer_offset(_In_ const ebpf_ring_buffer_t* ring)
+{
+    return ring->producer_offset % ring->length;
+}
+
+inline static size_t
+_ring_get_consumer_offset(_In_ const ebpf_ring_buffer_t* ring)
+{
+    return ring->consumer_offset % ring->length;
+}
+
+inline static void
+_ring_advance_producer_offset(_Inout_ ebpf_ring_buffer_t* ring, size_t length)
+{
+    ring->producer_offset += length;
+}
+
+inline static void
+_ring_advance_consumer_offset(_Inout_ ebpf_ring_buffer_t* ring, size_t length)
+{
+    ring->consumer_offset += length;
+}
+
+inline static _Ret_notnull_ ebpf_ring_buffer_record_t*
+_ring_record_at_offset(_In_ const ebpf_ring_buffer_t* ring, size_t offset)
+{
+    return (ebpf_ring_buffer_record_t*)&ring->shared_buffer[offset % ring->length];
+}
+
+inline static _Ret_notnull_ ebpf_ring_buffer_record_t*
+_ring_next_consumer_record(_In_ const ebpf_ring_buffer_t* ring)
+{
+    return _ring_record_at_offset(ring, _ring_get_consumer_offset(ring));
+}
+
+inline static _Ret_maybenull_ ebpf_ring_buffer_record_t*
+_ring_buffer_acquire_record(_Inout_ ebpf_ring_buffer_t* ring, size_t requested_length)
+{
+    ebpf_ring_buffer_record_t* record = NULL;
+    requested_length += EBPF_OFFSET_OF(ebpf_ring_buffer_record_t, data);
+    size_t remaining_space = ring->length - ring->producer_offset - ring->consumer_offset;
+
+    if (remaining_space > requested_length) {
+        record = _ring_record_at_offset(ring, _ring_get_producer_offset(ring));
+        _ring_advance_producer_offset(ring, requested_length);
+        record->header.length = (uint32_t)requested_length;
+        record->header.locked = 1;
+        record->header.discarded = 0;
+    }
+    return record;
+}
+
+ebpf_result_t
+ebpf_ring_buffer_create(_Outptr_ ebpf_ring_buffer_t** ring, size_t capacity)
+{
+    ebpf_result_t result;
+    ebpf_ring_buffer_t* local_ring_buffer = ebpf_epoch_allocate(sizeof(ebpf_ring_buffer_t));
+    if (!local_ring_buffer) {
+        result = EBPF_NO_MEMORY;
+        goto Error;
+    }
+
+    if ((capacity & ~(capacity - 1)) != capacity) {
+        result = EBPF_INVALID_ARGUMENT;
+        goto Error;
+    }
+
+    local_ring_buffer->length = capacity;
+
+    local_ring_buffer->ring_descriptor = ebpf_allocate_ring_buffer_memory(capacity);
+    if (!local_ring_buffer->ring_descriptor) {
+        result = EBPF_NO_MEMORY;
+        goto Error;
+    }
+    local_ring_buffer->shared_buffer = ebpf_ring_descriptor_get_base_address(local_ring_buffer->ring_descriptor);
+
+    *ring = local_ring_buffer;
+    local_ring_buffer = NULL;
+    return EBPF_SUCCESS;
+
+Error:
+    ebpf_ring_buffer_destroy(local_ring_buffer);
+    local_ring_buffer = NULL;
+    return result;
+}
+
+void
+ebpf_ring_buffer_destroy(_Frees_ptr_opt_ ebpf_ring_buffer_t* ring)
+{
+    if (ring) {
+        ebpf_free_ring_buffer_memory(ring->ring_descriptor);
+    }
+    ebpf_epoch_free(ring);
```

Comment:
since you need an if test anyway, it's ok to move 110 up inside the {} if you want